core/templates/components/graph-services/graph-layout.service.ts

```diff
@@ -547,64 +548,65 @@ export class StateGraphLayoutService {
         var targetWidth = link.target.width;
         var targetHeight = link.target.height;
 
-        var dx = targetx - sourcex;
-        var dy = targety - sourcey;
-
-        /* Fractional amount of truncation to be applied to the end of
-           each link. */
-        var startCutoff = (sourceWidth / 2) / Math.abs(dx);
-        var endCutoff = (targetWidth / 2) / Math.abs(dx);
-        if (dx === 0 || dy !== 0) {
-          startCutoff = (
-            (dx === 0) ? (sourceHeight / 2) / Math.abs(dy) :
-            Math.min(startCutoff, (sourceHeight / 2) / Math.abs(dy)));
-          endCutoff = (
-            (dx === 0) ? (targetHeight / 2) / Math.abs(dy) :
-            Math.min(endCutoff, (targetHeight / 2) / Math.abs(dy)));
-        }
+        if (targetx && sourcex && targety && sourcey && targetWidth &&
+          sourceWidth && targetHeight && sourceHeight) {
```

Comment:
Why is this needed?

Reply:
because all these values can be null in lines 55-58:
 ` yLabel: number | null;`
 ` xLabel: number | null;`
 ` height: number | null;`
 ` width: number | null;`

so as given at lines 537-540:
`  var sourcex = link.source.xLabel;`
`  var sourcey = link.source.yLabel;`
`  var targetx = link.target.xLabel;`
`  var targety = link.target.yLabel;`    
and at lines 547-550:
`  var sourceWidth = link.source.width;`
`  var sourceHeight = link.source.height;`
`  var targetWidth = link.target.width;`
`  var targetHeight = link.target.height;` 

these values (i.e, targetx, sourcex, targety, sourcey, targetWidth, sourceWidth, targetHeight, sourceHeight)
can also be `null`. Therefore to avoid  the following TS errors:

`core/templates/components/graph-services/graph-layout.service.ts(566,22): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(566,32): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(567,22): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(567,32): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(572,20): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(572,49): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(573,20): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(573,49): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(574,22): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(575,22): error TS2531: Object is possibly 'null'.`
`core/templates/components/graph-services/graph-layout.service.ts(576,20): error TS2531: Object is possibly 'null'.`

I have written this if-else statement. We can also put them inside ` Number()` as done at line 503.

.

Reply:
The best approach would again be to get rid of the `null`s, or at least make only one of the values indicate the the rest will be `null`. But if we keep the if it should look like this.
```suggestion
        if (
          target !== null && 
          sourcex !== null && 
          targety !== null && 
          sourcey !== null &&
          targetWidth  !== null && 
          sourceWidth !== null && 
          targetHeight !== null && 
          sourceHeight !== null 
        ) {
```